### ISSUE TITLE
Creating Morphing Animations with CSS clip-path

### ISSUE THREAD

**Issue by nirjan-dev** [nirjan-dev/learning-to-code#366]
[link to the post](https://blog.bitsrc.io/creating-morphing-animations-with-css-clip-path-3c3bf5e4335f)

Morphing is a technique where you transform one appearance into another. It’s a shape-shifting animation where a two-dimensional object, say an image or a figure or similar, turns into a different shape.

# Morphing in SVG

Most of the morphing effects, seen on the web today, comes in SVG. Vector graphics are all about coordinates. SVG is thus a natural starting point to create morphing effects.

# The two steps of creating the morphing effect
Creating a Morphing animation can, in rough terms, be divided into the following two activities:

   - Modeling
   - Animation

Modeling refers to the act of creating different vector patterns in the Morphing effect. I almost always start by creating the shapes in Inkscape, a vector editor. 

# Modeling the play/pause button

The animation in the demo consists of two separate shapes. What we’re looking at is an animation between a play symbol and a pause symbol. The play symbol comes as a triangle with rounded corners. The pause symbol is quite different; it consists of two parallel bars. As the two parts come in an uneven amount of shapes, two vs. one, we need to combine the two somehow to create the effect. The easiest way is to split the play symbol, as seen above, into two equal parts. Using two even symmetrical parts in animation is generally a good idea. There is something about symmetry. It usually creates, to the eye, a pleasing effect. Last but not least, there’s a slight rotation to account for the difference in the angle and placement.

Below is a figure of what the model of half the play symbol looks like in a vector editor.

![image](https://user-images.githubusercontent.com/29270715/81552700-a8d97f80-93a3-11ea-9400-18ffceb234c1.png)

Let’s zoom in a bit and have a look at the composition of the shape. Each half consists of 64 nodes each. Note the dense amount of nodes on the left side. The uneven placement of nodes here is deliberate. Nodes are more concentrated on the left side. This placement adds to the effect. The bottom corner will seemingly move when the triangle changes its shape. That is when three corners transform into four.

![image](https://user-images.githubusercontent.com/29270715/81552708-ae36ca00-93a3-11ea-8fd3-432e83bc055e.png)
Let’s have a look at what the model looks like in code:

```html
// Play/pause model
<svg ...>
  <circle ... /> /* red circle */
  <path d="..." /> /* Half play symbol */
  <path d="..." /> /* Half play symbol */
  <path d="..." /> /* Pause bar, left */
  <path d="..." /> /* Pause bar, right */
</svg>
```

# Making things move
Creating movement in the SVG space is, so far, a manual process. Here one needs to interpolate all the separate coordinates manually. One way to accomplish this is, for instance, with Javascript, using the requestAnimationFrame API call. With the callback function, one can calculate values for each frame. However, creating manual animations is cumbersome and takes time to set up. Another option is to use an animation framework. Doing so relieves you of some of the burdens to build the animation ground up. Either way, you’ll somehow still need to set up and initiate the animation programmatically. In short, an effort is necessary.

# What is this clip-path morphing?

oving to CSS makes it possible to leverage from the smart CSS transition engine. The engine practically does all of the heavy lifting. It covers the complete animation step and, which is important too, it does so natively. Native animations are accelerated and naturally performant.

Another thing we get for free is the complete interpolation of the full set of coordinates. The engine will dissect the collection and separately handle and interpolate each of the coordinates. The only thing we need to do is to provide the browser the starting shape and the end-state.

```css
.animated {
  transition: clip-path 500ms;
  clip-path: polygon(...); // start state
}.active .animated {
  clip-path: polygon(...); // end state
}
```

# Dev Tools FTW

As if the above was not enough, there’s also a great benefit using clipping paths in the Dev Tools. They provide inspection and direct manipulation of CSS clip-path rule. With the Dev Tools at your hands, you can test out animations directly in the roles editor. The inspector makes it even easier to iterate and try out different shapes.

# Moving to CSS space.
The conversion to CSS is an under-the-hood change. The idea is to create a substitute markup that produces the same visual output as the original markup. Below is an illustration in pseudo-code of what such a conversion could look like.

```html
/* From SVG */
<svg ...>
  <path d="..." />
</svg>/* To CSS */
<div style="clip-path: polygon(...)"></div>
```

# A glimpse into the future
There’s a supercharged upgrade around the corner. Namely, clip-path path(). Yes, I know, the name is even worse than clip-path morphing. However, this feature is the next generation SVG to CSS conversion. Once browsers fully support clip-path: path(), it’s going to be a breeze to go from SVG to CSS. 

# demo
[link to the demo](https://codepen.io/ainalem/pen/MWYrojE)

# TLDR;
 -   Use SVG to create the morphing start and end shapes
 -   Convert SVG paths to CSS clip-path polygons
 -   Use the Dev Tools to iterate and finetune the morphing effect.
 -   Trim the figures, easing and animation timing to perfect your animation
 -   Voila, you have created a brilliant morphing effect
 -   Clip-path: path() is going to be great
